proxy/src/connection.rs

```diff
@@ -187,49 +233,44 @@ impl AsyncWrite for Connection {
     }
 
     fn write_buf<B: Buf>(&mut self, buf: &mut B) -> Poll<usize, io::Error> {
-        use self::Connection::*;
-
-        match *self {
-            Plain(ref mut t) => t.write_buf(buf),
+        match self.io {
+            Io::Plain(ref mut t) => t.write_buf(buf),
         }
     }
 }
 
 impl Peek for Connection {
-    fn peek(&mut self, buf: &mut [u8]) -> io::Result<usize> {
-        use self::Connection::*;
-
-        match *self {
-            Plain(ref mut t) => t.peek(buf),
+    fn poll_peek(&mut self) -> Poll<usize, io::Error> {
+        if self.peek_buf.is_empty() {
+            self.peek_buf.reserve(8192);
```

Comment:
In the case of TLS 1024 is going to be more than sufficient in the vast majority of cases, however theoretically we might need to peek further than even 8192 bytes. If we're going to cap peeking at 8192 bytes, that's reasonable to do, but we should note that this limits, in theory but probably not in practice, our ability to do protocol detection.

In particular, if we cap the peeking at an arbitrary number like this then we'll need to have a way for the protocol detection logic to say "we didn't peek enough so we can't detect the protocol."

Reply:
I chose 8192 as its a commonish max size of a packet, while not being too big. I figured we'd need to adjust a little when TLS is put in place, since I don't know exactly what that'd look. I could also just drop this down to 1kb or smaller too, but I don't have great reasons for doing so that are better than just trying to fit the first read into a single buffer/syscall.